File: 13. Roman to Integer/solution.cpp
/*
    Runtime: 4 ms, faster than 99.74% of C++ online submissions for Roman to Integer.
    Memory Usage: 8.6 MB, less than 80.30% of C++ online submissions for Roman to Integer.
*/

class Solution {
public:
    int romanToInt(string s) {
        int res = 0;
        char tmp = '\0';
        for(int i = s.size() - 1; i >= 0; --i){
        	if(s[i] == 'I'){
        		if(tmp != 'V' && tmp != 'X'){
        			res++;
        		} else {
        			res--;
        		}
        	} else if(s[i] == 'V'){
        		res += 5;
        		tmp = 'V';
        	} else if(s[i] == 'X'){
        		if(tmp != 'L' && tmp != 'C'){
        			res += 10;
        			tmp = 'X';
        		} else {
        			res -= 10;
        		}
        	} else if(s[i] == 'L'){
        		res += 50;
        		tmp = 'L';
        	} else if(s[i] == 'C'){
        		if (tmp != 'D' && tmp != 'M'){
					res += 100;
					tmp = 'C';
        		} else {
        			res -= 100;
        		}
        	} else if(s[i] == 'D'){
        		res += 500;
        		tmp = 'D';
        	} else if(s[i] == 'M'){
        		res += 1000;
        		tmp = 'M';
        	}
        }
        return res;
    }
};

File: 13. Roman to Integer/solution.py
#   Runtime: 48 ms, faster than 99.57% of Python3 online submissions for Roman to Integer.
#   Memory Usage: 13.3 MB, less than 49.06% of Python3 online submissions for Roman to Integer.

class Solution:
    def romanToInt(self, s: str) -> int:
        romain_int = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        }
        s = s.replace("IV", "IIII").replace("IX", "VIIII")
        s = s.replace("XL", "XXXX").replace("XC", "LXXXX")
        s = s.replace("CD", "CCCC").replace("CM", "DCCCC")

        return sum([romain_int[c] for c in s])
